mypy/checkexpr.py

```diff
@@ -823,20 +823,36 @@ def infer_function_type_arguments_using_context(
         # valid results.
         erased_ctx = replace_meta_vars(ctx, ErasedType())
         ret_type = callable.ret_type
-        if isinstance(ret_type, TypeVarType):
-            if ret_type.values or (not isinstance(ctx, Instance) or
-                                   not ctx.args):
-                # The return type is a type variable. If it has values, we can't easily restrict
-                # type inference to conform to the valid values. If it's unrestricted, we could
-                # infer a too general type for the type variable if we use context, and this could
-                # result in confusing and spurious type errors elsewhere.
-                #
-                # Give up and just use function arguments for type inference. As an exception,
-                # if the context is a generic instance type, actually use it as context, as
-                # this *seems* to usually be the reasonable thing to do.
-                #
-                # See also github issues #462 and #360.
-                ret_type = NoneTyp()
+        if mypy.checker.is_optional(ret_type) and mypy.checker.is_optional(ctx):
+            # If both the context and the return type are optional, unwrap the optional,
+            # since in 99% cases this is what a user expects. In other words, we replace
+            #     Optional[T] <: Optional[int]
+            # with
+            #     T <: int
+            # while the former would infer T <: Optional[int].
+            ret_type = mypy.checker.remove_optional(ret_type)
+            erased_ctx = mypy.checker.remove_optional(erased_ctx)
+            #
+            # TODO: Instead of this hack and the one below, we need to use outer and
+            # inner contexts at the same time. This is however not easy because of two
+            # reasons:
+            #   * We need to support constraints like [1 <: 2, 2 <: X], i.e. with variables
+            #     on both sides. (This is not too hard.)
+            #   * We need to update all the inference "infrastructure", so that all
+            #     variables in an expression are inferred at the same time.
+            #     (And this is hard, also we need to be careful with lambdas that require
+            #     two passes.)
+        if isinstance(ret_type, TypeVarType) and not (isinstance(ctx, Instance) and ctx.args):
+            # Another special case: the return type is a type variable. If it's unrestricted,
+            # we could infer a too general type for the type variable if we use context,
+            # and this could result in confusing and spurious type errors elsewhere.
+            #
+            # Give up and just use function arguments for type inference. As an exception,
+            # if the context is a generic instance type, actually use it as context, as
+            # this *seems* to usually be the reasonable thing to do.
```

Comment:
TBH I am not sure this is better. Currently, we ignore the return context for cases like `T <: int` but use it for `T <: Iterable[int]`. So the only way how only using `T <: List[int]` will be better, is because invariant containers statistically have less subtypes. As I understand the original problem this solves is like this:
```python
def f(x: List[T]) -> T: ...
y: object = f([1, 2, 3])  # Incompatible argument type "List[int]", expected "List[object]"
```
In other words the problem appears when the return type appears as a variable in invariant context in one of argument types.

However, if you prefer to only use invariant ones, I can update this.

Reply:
What I meant is that if the context is, say, `List[object]`, `List[int]` wouldn't be valid due to invariance, so we are more likely to want to use the surrounding type context and not rely on arguments only. However, if the context is `Iterable[object]`, `Iterable[int]` would be okay, due co covariance, so we ignore the context. Whether there are many subtypes or not doesn't seem significant?

Reply:
OK, this is actually exactly what I meant :-) `Iterable[object]` has more subtypes, such as `Iterable[int]`. While `List[object]` doesn't have such, so it is typically better to still use it as a context. OK, I will modify the special case to reflect this.